Library/Homebrew/test/test_cmd_audit.rb

```diff
@@ -0,0 +1,251 @@
+    assert_equal ["File should end with a newline"], fa.problems
+  end
+
+  def test_audit_file_not_strict_no_issue
+    fa = formula_auditor "foo", <<-EOS.undent
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+        homepage "http://example.com"
+      end
+    EOS
+    fa.audit_file
+    assert_equal [], fa.problems
+  end
+
+  def test_audit_file_strict_ordering_issue
+    fa = formula_auditor "foo", <<-EOS.undent, :strict => true
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+        homepage "http://example.com"
+      end
+    EOS
+    fa.audit_file
+    assert_equal ["`homepage` (line 3) should be put before `url` (line 2)"],
+      fa.problems
+  end
+
+  def test_audit_file_strict_url_outside_of_stable_block
+    fa = formula_auditor "foo", <<-EOS.undent, :strict => true
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+        stable do
+          # stuff
+        end
+      end
+    EOS
+    fa.audit_file
+    assert_equal ["`url` should be put inside `stable block`"], fa.problems
+  end
+
+  def test_audit_file_strict_head_and_head_do
+    fa = formula_auditor "foo", <<-EOS.undent, :strict => true
+      class Foo < Formula
+        head "http://example.com/foo.git"
+        head do
+          # stuff
+        end
+      end
+    EOS
+    fa.audit_file
+    assert_equal ["Should not have both `head` and `head do`"], fa.problems
+  end
+
+  def test_audit_file_strict_bottle_and_bottle_do
+    fa = formula_auditor "foo", <<-EOS.undent, :strict => true
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+        bottle do
+          # bottles go here
+        end
+        bottle :unneeded
+      end
+    EOS
+    fa.audit_file
+    assert_equal ["Should not have `bottle :unneeded/:disable` and `bottle do`"],
+      fa.problems
+  end
+
+  def test_audit_class_no_test
+    fa = formula_auditor "foo", <<-EOS.undent
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+      end
+    EOS
+    fa.audit_class
+    assert_equal [], fa.problems
+
+    fa = formula_auditor "foo", <<-EOS.undent, :strict => true
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+      end
+    EOS
+    fa.audit_class
+    assert_equal ["A `test do` test block should be added"], fa.problems
+  end
+
+  def test_audit_class_github_gist_formula
+    fa = formula_auditor "foo", <<-EOS.undent
+      class Foo < GithubGistFormula
+        url "http://example.com/foo-1.0.tgz"
+      end
+    EOS
+    fa.audit_class
+    assert_equal ["GithubGistFormula is deprecated, use Formula instead"],
+      fa.problems
+  end
+
+  def test_audit_class_script_file_formula
+    fa = formula_auditor "foo", <<-EOS.undent
+      class Foo < ScriptFileFormula
+        url "http://example.com/foo-1.0.tgz"
+      end
+    EOS
+    fa.audit_class
+    assert_equal ["ScriptFileFormula is deprecated, use Formula instead"],
+      fa.problems
```

Comment:
I think because both `test/test_formula_installer.rb` and `test/test_formula_installer_bottle.rb` have a `require "compat/formula_specialties"` in them. If they ran before this test, then those classes are available. (This will lead to random failures if test order is different.)